aptos-move/move-examples/post_mint_reveal_nft/sources/minting.move

```diff
@@ -0,0 +1,861 @@
+                signer::address_of(&resource_signer),
+                1,
+                0,
+                create_token_mutability_config(
+                    &vector<bool>[ false, false, false, false, true ]
+                ),
+                vector<String>[string::utf8(BURNABLE_BY_OWNER)],
+                vector<vector<u8>>[bcs::to_bytes<bool>(&true)],
+                vector<String>[string::utf8(b"bool")],
+            );
+
+            move_to(&resource_signer, SourceToken { token_data_id: source_token_data_id });
+        };
+
+        // if this is the whitelist minting time
+        if (now > whitelist_mint_config.whitelist_minting_start_time && now < whitelist_mint_config.whitelist_minting_end_time) {
+            let claimer_addr = signer::address_of(nft_claimer);
+            assert!(bucket_table::contains(&whitelist_mint_config.whitelisted_address, &claimer_addr), error::permission_denied(EACCOUNT_NOT_WHITELISTED));
+            let remaining_mint_allowed = bucket_table::borrow_mut(&mut whitelist_mint_config.whitelisted_address, claimer_addr);
+            assert!(*remaining_mint_allowed >= amount, error::invalid_argument(EAMOUNT_EXCEEDS_MINTS_ALLOWED));
+            while (amount > 0) {
+                mint_source_certificate_internal(nft_claimer, whitelist_mint_config.whitelist_mint_price);
+                amount = amount - 1;
+                *remaining_mint_allowed = *remaining_mint_allowed - 1;
+            };
+        } else {
+            while (amount > 0) {
+                mint_source_certificate_internal(nft_claimer, public_mint_config.public_mint_price);
+                amount = amount - 1;
+            };
+        };
+    }
+
+    // This will burn the source certificate token and put a destination token in the nft_claimer's TokenStore.
+    entry fun exchange(nft_claimer: &signer, token_id: TokenId) acquires NFTMintConfig, CollectionConfig, PublicMintConfig {
+        assert!(exists<NFTMintConfig>(@post_mint_reveal_nft) && exists<CollectionConfig>(@post_mint_reveal_nft) && exists<PublicMintConfig>(@post_mint_reveal_nft), error::permission_denied(ECONFIG_NOT_INITIALIZED));
+
+        let public_mint_config = borrow_global<PublicMintConfig>(@post_mint_reveal_nft);
+        assert!(timestamp::now_seconds() > public_mint_config.public_minting_end_time, error::permission_denied(ECANNOT_EXCHANGE_BEFORE_MINTING_ENDS));
+        assert!(token::balance_of(signer::address_of(nft_claimer), token_id) > 0, error::invalid_argument(ETOKEN_ID_NOT_FOUND));
+
+        let now = timestamp::now_microseconds();
+        let nft_mint_config = borrow_global_mut<NFTMintConfig>(@post_mint_reveal_nft);
+        let token_config = borrow_global_mut<CollectionConfig>(@post_mint_reveal_nft);
+
+        // Assert there's still some token uris in the vector.
+        assert!(vector::length(&token_config.token_uri) > 0, error::resource_exhausted(ENO_ENOUGH_TOKENS_LEFT));
+
+        let index = now % vector::length(&token_config.token_uri);
```

Comment:
This is an odd formula. Can you elaborate what is this for and why it's done this way?

Reply:
this is for pseudo randomizing what token the user is getting.
here we're assuming the `timestamp::now_microseconds` value and the `vector::length(&token_config.token_uri)` value (the vector::length will be diff after each `exchange` transaction) are both unpredictable.

Reply:
Timestamp is subject to some degree of manipulation by the proposer. They can tweak the timestamp within a narrow range and still pass consensus.

Reply:
yeah, but this is just for pseudo-randomization... also if they tweak the timestamp, the vector length could change, since there could be other `exchange()` transactions before that transaction gets in. In the `exchange()` function we remove the destination token once it's minted, so the length of the `token_uri` vector reduces by 1 in every successful `exchange()` call.